Initial infrastructure for the C-Find command.

leonardorame opened the pull request:
Hi @jjedele I'm building the basic infrastructure for the C-Find command, could you take a look at this?

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -158,9 +158,24 @@ defmodule DicomNet.Association do
-    msg = {:dicom, %{operation: :cstore, dataset: ds}}
-    send(event_listener, msg)
-    response_ds = handle_cstore(command, ds)
+    commandField = command[256].values 
```
I would stay with using `Dataset.fetch!(:Keyword)`. That's much easier to read and doesn't force DataSet to always remain a simple map.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -158,9 +158,24 @@ defmodule DicomNet.Association do
+    commandField = command[256].values 
+
+    response_ds = case commandField do
+        [1] -> IO.inspect("C-Store")
```
Would stay with hex literals in this code, as this is how the DICOM standard describes it as well. Easier to map then.

leonardorame replied:
Yes, I used [1] and [256] because I don't know if I could use, for example 0x0001 directly, or [0x0001]. Which is the recommended way?

jjedele replied:
What I mean is the following:

On line 161 you access a named field in the command dataset, which you best do with:
```
command_field = DataSet.fetch!(:CommandField)
```

Then:
```
case command_field do
  0x01 -> ...
  0x20 -> ...
```

This way I can compare the code directly to the DICOM standard (in this case [here](https://dicom.nema.org/dicom/2013/output/chtml/part07/sect_9.3.html#table_9.3-3)), and see that it's correct without converting the number in my head.

leonardorame replied:
Thanks!, that's what I needed.

jjedele reviewed (commented):
Good start.

I think the most interesting parts for C-FIND will be:

1. The bidirectional API. C-Store just receives Datasets and forwards them to a handler to do something with them. C-FIND and C-GET will need to forward the query to a handler, and then receive one or more Datasets as answer to return them to the requester.
2. You will need to implement Dataset serialization before. So far I think I did only the minimum necessary part to be able to send responses back.

leonardorame commented:
There is some progress in my implementation. At least the findscu command gets results, but I cannot yet return a list of responses, here's what I get:

```
D: ====================== BEGIN A-ASSOCIATE-AC =====================                                                                                                                          
D: Our Implementation Class UID:      1.2.276.0.7230010.3.0.3.6.2                                                                                                                             
D: Our Implementation Version Name:   OFFIS_DCMTK_362                                                                                                                                         
D: Their Implementation Class UID:    1.2.276.0.7230010.3.0.3.6.2                                                                                                                             
D: Their Implementation Version Name: OFFIS_DCMTK_362                                                                                                                                         
D: Application Context Name:    1.2.840.10008.3.1.1.1                                                                                                                                         
D: Calling Application Name:    FINDSCU                                                                                                                                                       
D: Called Application Name:     ANY-SCP                                                                                                                                                       
D: Responding Application Name: ANY-SCP                                                                                                                                                       
D: Our Max PDU Receive Size:    16384                                                                                                                                                         
D: Their Max PDU Receive Size:  16384                                                                                                                                                         
D: Presentation Contexts:                                                                                                                                                                     
D:   Context ID:        1 (Accepted)                                                                                                                                                          
D:     Abstract Syntax: =FINDModalityWorklistInformationModel                                                                                                                                 
D:     Proposed SCP/SCU Role: Default                                                                                                                                                         
D:     Accepted SCP/SCU Role: Default                                                                                                                                                         
D:     Accepted Transfer Syntax: =LittleEndianExplicit                                                                                                                                        
D: Requested Extended Negotiation: none                                                                                                                                                       
D: Accepted Extended Negotiation:  none  
D: Requested User Identity Negotiation: none
D: User Identity Negotiation Response:  none
D: ======================= END A-ASSOCIATE-AC ======================
I: Association Accepted (Max Send PDV: 16372)
I: Sending Find Request
D: ===================== OUTGOING DIMSE MESSAGE ====================
D: Message Type                  : C-FIND RQ
D: Presentation Context ID       : 1
D: Message ID                    : 1
D: Affected SOP Class UID        : FINDModalityWorklistInformationModel
D: Data Set                      : present
D: Priority                      : low
D: ======================= END DIMSE MESSAGE =======================
I: Request Identifiers:
I: 
I: # Dicom-Data-Set
I: # Used TransferSyntax: Little Endian Explicit
I: (0008,0050) SH (no value available)                     #   0, 0 AccessionNumber
I: (0008,0052) CS [STUDY]                                  #   6, 1 QueryRetrieveLevel
I: 
D: DcmDataset::read() TransferSyntax="Little Endian Implicit"
W: DcmItem: Element (0000,0000) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0800) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0900) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0000) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0800) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0900) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0008,0050) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0000) found twice in one data set or item, ignoring second entry
W: DcmItem: Dataset not in ascending tag order, at element (0000,0002)
W: DcmItem: Dataset not in ascending tag order, at element (0000,0100)
W: DcmItem: Dataset not in ascending tag order, at element (0000,0120)
W: DcmItem: Element (0000,0800) found twice in one data set or item, ignoring second entry
W: DcmItem: Element (0000,0900) found twice in one data set or item, ignoring second entry
I: Received Final Find Response
D: ===================== INCOMING DIMSE MESSAGE ====================
D: Message Type                  : C-FIND RSP
D: Message ID Being Responded To : 1
D: Affected SOP Class UID        : FINDModalityWorklistInformationModel
D: Data Set                      : none
D: DIMSE Status                  : 0x0000: Success
D: ======================= END DIMSE MESSAGE =======================
D: Status Detail:
D: 
D: # Dicom-Data-Set
D: # Used TransferSyntax: Little Endian Implicit
D: (0008,0050) SH [123]                                   #   4, 1 AccessionNumber
D: (0008,0052) CS [STUDY]                                 #   6, 1 QueryRetrieveLevel
D: 
I: Releasing Association
```

I'm playing with creating a list of datasets, then serialize each and concatenate. But there's something else to do, and I don´t know what it is.

leonardorame commented:
What I still don't get, is all c-find-rsp must be returned in one big buffer or there's some intereaction between the scu and the scp, one c-find-rsp for each found item.

jjedele commented:
Hey - getting this in a state where the client understands the response is a good start!

I haven't done a lot with C-FIND yet either, but I understand the [protocol](https://dicom.nema.org/medical/dicom/current/output/chtml/part07/sect_9.3.2.4.html) as follows:

1. You send a [C-FIND-RQ ](https://dicom.nema.org/medical/dicom/current/output/chtml/part07/sect_9.3.2.html) to the SCP with a query dataset which will be matched against all DICOM instances it knows
2. For each matching instance, the SCP answers with a separate [C-FIND-RSP](https://dicom.nema.org/medical/dicom/current/output/chtml/part07/sect_9.3.2.2.html). All responses shall have the [Status](https://dicom.nema.org/medical/dicom/current/output/chtml/part07/chapter_C.html) `Pending`; and the final one  `Success` to indicate to the SCU that the results are complete.

What might also be helpful is installing Wireshark and listening in to some C-FIND interaction between other tools. Seeing it in action often makes things clearer.

P.S.: I'll not be much on my laptop from tomorrow on over the weekend, so I'll probably won't get to answer in the meantime.

leonardorame commented:
I'm debugging the TCP trace with Wireshark and found that there's no ACK after every RSP.

This is a correct trace (against a real PACS):
![image](https://github.com/user-attachments/assets/d81967f6-b22a-4c1c-8c5d-dda9288c51e0)
As you can see, after every `C-FIND-RSP ID=1` there's an `[ACK]` followed by an `[PSH, ACK]`, and after that the `C-FIND-RSP-DATA`.

Now, this is the trace for my implementation:
![image](https://github.com/user-attachments/assets/e0ac80f9-ed71-487b-a2db-94d97070fb7b)
Here there's no `[ACK]` nor `[PSH, ACK]` after the `C-FIND-RSP`, also there's no `C-FIND-RSP-DATA`.

I think the issue is we are not recv'ing the response after every `:gen_tcp.send`.  Shouldn't the process be send->recv->send->recv and so on?

Here's my latest `handle_cfind` code. After every serialized part (header, rsp1, ..., rspN, tail) I try to `:gen_tcp.send` to interact with the C-FIND SCU, but there's something missing I cannot solve yet.

```elixir
defp handle_cfind(socket, command, data_set, presentation_context_id) do 
    qr_level = DataSet.fetch!(data_set, :QueryRetrieveLevel)
    asci_de = DataSet.fetch!(command, :AffectedSOPClassUID)
    mid_de = DataSet.fetch!(command, :MessageID)

    response_head =
      Dicom.DataSet.from_keyword_list(
        AffectedSOPClassUID: DataElement.value(asci_de),
        CommandField: 0x8020,
        MessageIDBeingRespondedTo: DataElement.value(mid_de),
        CommandDataSetType: 0x0001, # This field shall be set to the value of 0101H (Null) if no Data Set is present; any other value indicates a Data Set is included in the Message.
        Status: 0xff01 # 0x0000 Success, 0xff00 Pending
      )

    response = Dicom.BinaryFormat.serialize_command_data_set(response_head)
      |> Pdu.new_data_pdu(presentation_context_id)
      |> Pdu.serialize()

    :gen_tcp.send(socket, response)
    IO.inspect("--1--")
    send(self(), :read_data)
    response_rsp1 =                                                                                                                                                                           
      Dicom.DataSet.from_keyword_list(                                                                                                                                                        
        AccessionNumber: "aaaa"                                                                                                                                                               
      )                                                                                                                                                                                       
                                                                                                                                                                                              
    response = Dicom.BinaryFormat.serialize_command_data_set(response_rsp1)                                                                                                                   
      |> Pdu.new_data_pdu(presentation_context_id)                                                                                                                                            
      |> Pdu.serialize()                                                                                                                                                                      
                                                                                                                                                                                              
    :gen_tcp.send(socket, response)                                                                                                                                                           
    IO.inspect("--2--")                                                                                                                                                                       
    send(self(), :read_data)                                                                                                                                                                  
                                                                                                                                                                                              
    response_tail =                                                                                                                                                                           
      Dicom.DataSet.from_keyword_list(
        AffectedSOPClassUID: DataElement.value(asci_de),
        CommandField: 0x8020,
        MessageIDBeingRespondedTo: DataElement.value(mid_de),
        QueryRetrieveLevel: DataElement.value(qr_level),
        CommandDataSetType: 0x0101, # This field shall be set to the value of 0101H (Null) if no Data Set is present; any other value indicates a Data Set is included in the Message.
        Status: 0x0000  # 0xff0 Success, 0xff00 Pending
      )

    response = Dicom.BinaryFormat.serialize_command_data_set(response_tail)
      |> Pdu.new_data_pdu(presentation_context_id)
      |> Pdu.serialize()

    :gen_tcp.send(socket, response)
    send(self(), :read_data)

    response_tail
end
```

jjedele commented:
Hey,

still currently not much time to look into this more, but a couple of high-level pointers:

1. You can probably ignore the ACK and PSH messages for now. That's TCP-level stuff, and we can assume the network stack is doing this correctly. They might look a bit different depending on how the communication happens in detail.
2. In your trace, note the ABORT going from your client (FINDSCU) to the server (ANY-SCP). This most likely means that the findscu is not happy with your first response and thus aborts the query. Have you activated the debug mode (`-d`) to get more information about what's happening?
3. The good thing is Wireshark recognizes your response as a C-FIND-RSP, so it's probably not completely wrong. But in general you might want to try comparing packets byte-by-byte in small examples. Even little problems (header length bytes or something) could make the communication fail.
4. I'm not sure why not receiving from the socket after every send should be a problem, the sockets are fully bidirectional. If we don't expect any data we need in there, we also shouldn't need to read from them.

leonardorame commented:
_The good thing is Wireshark recognizes your response as a C-FIND-RSP, so it's probably not completely wrong. But in general you might want to try comparing packets byte-by-byte in small examples. **Even little problems (header length bytes or something)** could make the communication fail_.

Hi @jjedele, yes, I'm sure the problem is there, but I don't know how to proceed here. I'll need your help.

jjedele commented:
In case you didn't do it yet, the first step would be enabling debug mode
in findscu to find out why it aborts the connection with your server code.

On Mon, Feb 10, 2025, 21:54 Leonardo M. Ramé ***@***.***>
wrote:

> *The good thing is Wireshark recognizes your response as a C-FIND-RSP, so
> it's probably not completely wrong. But in general you might want to try
> comparing packets byte-by-byte in small examples. Even little problems
> (header length bytes or something) could make the communication fail*.
>
> Hi @jjedele <https://github.com/jjedele>, yes, I'm sure the problem is
> there, but I don't know how to proceed here. I'll need your help.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jjedele/dicom.ex/pull/10#issuecomment-2649212668>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF6MZOWNZ6SO4ADOPENOIT2PEGXZAVCNFSM6AAAAABWQZCDQCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBZGIYTENRWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

leonardorame commented:
Yes, I mostly always use it with debug mode.

jjedele commented:
Ok, can you then please attach a full log of it running against the latest
version of your code? Especially interesting are all warnings and errors of
course.

On Mon, Feb 10, 2025, 22:08 Leonardo M. Ramé ***@***.***>
wrote:

> Yes, I mostly always use it with debug mode.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jjedele/dicom.ex/pull/10#issuecomment-2649243377>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF6MZJNTEEHCNDGFTFSNN32PEIN5AVCNFSM6AAAAABWQZCDQCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBZGI2DGMZXG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

leonardorame commented:
Yes of course, here it is:

```
D: $dcmtk: findscu v3.6.2 2017-07-14 $                                                                                                                                                  [57/137]
D:                                                                                                                                                                                              
D: DcmDataDictionary: Loading file: /home/leonardo/desarrollo/informemedico/dicomscp/config/dicom.dic                                                                                           
D: Request Parameters:                                                                                                                                                                          
D: ====================== BEGIN A-ASSOCIATE-RQ =====================                                                                                                                            
D: Our Implementation Class UID:      1.2.276.0.7230010.3.0.3.6.2                                                                                                                               
D: Our Implementation Version Name:   OFFIS_DCMTK_362                                                                                                                                           
D: Their Implementation Class UID:                                                                                                                                                              
D: Their Implementation Version Name:                                                                                                                                                           
D: Application Context Name:    1.2.840.10008.3.1.1.1                                                                                                                                           
D: Calling Application Name:    FINDSCU                                                                                                                                                         
D: Called Application Name:     ANY-SCP                                                                                                                                                         
D: Responding Application Name: ANY-SCP                                                                                                                                                         
D: Our Max PDU Receive Size:    16384
D: Their Max PDU Receive Size:  0
D: Presentation Contexts:           
D:   Context ID:        1 (Proposed)                                                            
D:     Abstract Syntax: =FINDModalityWorklistInformationModel
D:     Proposed SCP/SCU Role: Default
D:     Proposed Transfer Syntax(es):                                                            
D:       =LittleEndianExplicit         
D:       =BigEndianExplicit            
D:       =LittleEndianImplicit              
D: Requested Extended Negotiation: none     
D: Accepted Extended Negotiation:  none                                                         
D: Requested User Identity Negotiation: none 
D: User Identity Negotiation Response:  none
D: ======================= END A-ASSOCIATE-RQ ======================
I: Requesting Association                   
D: setting network send timeout to 60 seconds
D: setting network receive timeout to 60 seconds 
D: Constructing Associate RQ PDU                                                                
D: PDU Type: Associate Accept, PDU Length: 186 + 6 bytes PDU header
D:   02  00  00  00  00  ba  00  01  00  00  41  4e  59  2d  53  43
D:   50  20  20  20  20  20  20  20  20  20  46  49  4e  44  53  43 
D:   55  20  20  20  20  20  20  20  20  20  00  00  00  00  00  00
D:   00  00  00  00  00  00  00  00  00  00  00  00  00  00  00  00
D:   00  00  00  00  00  00  00  00  00  00  10  00  00  15  31  2e
D:   32  2e  38  34  30  2e  31  30  30  30  38  2e  33  2e  31  2e
D:   31  2e  31  21  00  00  1b  01  00  00  00  40  00  00  13  31
D:   2e  32  2e  38  34  30  2e  31  30  30  30  38  2e  31  2e  32
D:   2e  31  50  00  00  3a  52  00  00  1b  31  2e  32  2e  32  37
D:   36  2e  30  2e  37  32  33  30  30  31  30  2e  33  2e  30  2e
D:   33  2e  36  2e  32  55  00  00  0f  4f  46  46  49  53  5f  44
D:   43  4d  54  4b  5f  33  36  32  51  00  00  04  00  00  40  00
D: 
D: Parsing an A-ASSOCIATE PDU
D: Association Parameters Negotiated:
D: ====================== BEGIN A-ASSOCIATE-AC =====================
D: Our Implementation Class UID:      1.2.276.0.7230010.3.0.3.6.2
D: Our Implementation Version Name:   OFFIS_DCMTK_362
D: Their Implementation Class UID:    1.2.276.0.7230010.3.0.3.6.2
D: Their Implementation Version Name: OFFIS_DCMTK_362
D: Application Context Name:    1.2.840.10008.3.1.1.1
D: Calling Application Name:    FINDSCU
D: Called Application Name:     ANY-SCP
D: Responding Application Name: ANY-SCP
D: Our Max PDU Receive Size:    16384
D: Their Max PDU Receive Size:  16384
D: Presentation Contexts:
D:   Context ID:        1 (Accepted)
D:     Abstract Syntax: =FINDModalityWorklistInformationModel
D:     Proposed SCP/SCU Role: Default
D:     Accepted SCP/SCU Role: Default
D:     Accepted Transfer Syntax: =LittleEndianExplicit
D: Requested Extended Negotiation: none
D: Accepted Extended Negotiation:  none
D: Requested User Identity Negotiation: none
D: User Identity Negotiation Response:  none
D: ======================= END A-ASSOCIATE-AC ======================
I: Association Accepted (Max Send PDV: 16372)
I: Sending Find Request
D: ===================== OUTGOING DIMSE MESSAGE ====================
D: Message Type                  : C-FIND RQ
D: Presentation Context ID       : 1
D: Message ID                    : 1
D: Affected SOP Class UID        : FINDModalityWorklistInformationModel
D: Data Set                      : present
D: Priority                      : low
D: ======================= END DIMSE MESSAGE =======================
I: Request Identifiers:
I: 
I: # Dicom-Data-Set
I: # Used TransferSyntax: Little Endian Explicit
I: (0008,0050) SH (no value available)                     #   0, 0 AccessionNumber
I: (0008,0052) CS [STUDY]                                  #   6, 1 QueryRetrieveLevel
I: 
D: DcmDataset::read() TransferSyntax="Little Endian Implicit"
E: Find Failed, query keys:
E: 
E: # Dicom-File-Format
E: 
E: # Dicom-Meta-Information-Header
E: # Used TransferSyntax: Little Endian Explicit
E: 
E: # Dicom-Data-Set
E: # Used TransferSyntax: Little Endian Explicit
E: (0008,0050) SH (no value available)                     #   0, 0 AccessionNumber
E: (0008,0052) CS [STUDY ]                                 #   6, 1 QueryRetrieveLevel
E: 
E: 0006:020f DIMSE Unexpected PDV type
E: Find SCU Failed: 0006:020f DIMSE Unexpected PDV type
I: Aborting Association
```

jjedele commented:
Hey,

so in general, when working on this level of DICOM, the reference you need to have at hand is the [Upper Layer protocol definition](https://dicom.nema.org/dicom/2013/output/chtml/part08/sect_9.3.html) (and [Annex E](https://dicom.nema.org/dicom/2013/output/chtml/part08/chapter_E.html)). This defines byte-by-byte how the messages must look like.

As you can see in the trace of findscu, it aborts because it doesn't understand the [PDV](https://dicom.nema.org/dicom/2013/output/chtml/part08/sect_9.3.html#sect_9.3.5.1) type (presentation-data-value). These get their name because they are encoded using the transfer syntax of the negotiated presentation context.

Now one thing I see, is that you encode the data part of your response also using `BinaryFormat.serialize_command_data_set`, which is a convenience method always encoding using ImplicitVRLittleEndian TransferSyntax and adding the CommandGroupLength attribute. Your findscu logs however show, that the tools agreed on using `ExplicitVRLittleEndian`. If it then tries to parse implicit VR data as explicit VR data, that would be one explanation of why it breaks with "invalid PDV".

The proper solution here is to implement a proper `BinaryFormat.serialize` method, which takes the encoding options (VR explicitness, endianness) from the negotiated transfer syntax and serializes the response dataset accordingly (see [here](https://github.com/jjedele/dicom.ex/blob/main/lib/dicom_net/association.ex#L154-L160) for the parsing equivalent in the current c-store).

To find out if this is the (only) problem, you can also try to enforce ImplicitVRLittleEndian Transfer Syntax being used (according to the [findscu manual](https://support.dcmtk.org/docs/findscu.html), `-xi -xe` should the simplest way of doing this. (It could be that it still has an issue with the CommandGroupLength attribute which shouldn't be in the data part. But I would guess it'll just be ignored since it's standard DICOM).

leonardorame commented:
Hi @jjedele, I'm trying to understand how to build an Explicit Little Endian serializer. As far as I've read the only difference with Implicit Little Endian is the VR must be added to the encoded attributes.

I tried add the serializer by simply adding the VR as this:
`
group <> element <> vr <> value_length <> value`

Note that I just added the vr. 

For this I simply duplicated you `serialize_data_element` function and added the param `explicit: true`, and created a convenience function converts each VR into its byte representation.

```
def vr_to_hex(vr_atom) do
    case vr_atom do
      :US -> <<0x55, 0x53>>
      :UL -> <<0x55, 0x4C>>
      :UI -> <<0x55, 0x49>>
      :LO -> <<0x4C, 0x4F>>
      :SH -> <<0x53, 0x48>>
      :CS -> <<0x43, 0x54>>
    end
  end

  def serialize_data_element(data_element, endianness: endianness, explicit: true) do
    group = serialize_u16(data_element.group_number, endianness)
    element = serialize_u16(data_element.element_number, endianness)

    # TODO cannot handle VM > 1
    value =
      case data_element.vr do
        :US -> serialize_u16(data_element |> DataElement.value(), endianness)
        :UL -> serialize_u32(data_element |> DataElement.value(), endianness)
        :UI -> serialize_uid(data_element |> DataElement.value(), endianness)
        :LO -> serialize_u16(data_element |> DataElement.value(), endianness)
        :SH -> serialize_sh(data_element |> DataElement.value(), endianness)
        :CS -> serialize_cs(data_element |> DataElement.value(), endianness)
      end

    value_length = serialize_u32(byte_size(value), endianness)

    group <> element <> vr_to_hex(data_element.vr) <> value_length <> value
  end
```

This doesn't work, but I would like to know if I'm going into the right direction.

jjedele commented:
It's a bit more complicated than this. The encoding behaves differently depending on VR, because the length of their contained values can differ quite a lot. See [section 7.1.2](https://dicom.nema.org/dicom/2013/output/chtml/part05/chapter_7.html#sect_7.1.2).

I already have [fairly good unit tests](https://github.com/jjedele/dicom.ex/blob/main/test/dicom/binary_format_test.exs) for the parsing, there you have samples for each VR and all encoding options. I should also be pretty easy to refactor these unit tests so that they can easily cover both parsing and serialization, which I was planning to do anyways.

But before going to deep into this: Did you try if it works if you force the SCU to use implicit VR?

leonardorame commented:
Great news!, finally I was able to make it work.

In binary_format.exe I added this function (very similar to `serialize_command_data_set`, but using `explicit: true` and removing the command group part):

```elixir
  def serialize_data_data_set(data_set) do
    serialization_options = [endianness: :little, explicit: true]
    serialized =
      for {_tag, de} <- data_set do
        Dicom.BinaryFormat.serialize_data_element(de, serialization_options)                          
      end
      |> Enum.into(<<>>)
    serialized
  end
```

Then, as the serialization options uses explicit: true I created a new Dicom.BinaryFormat.serialize_data_element function as this:

```elixir
  def vr_to_hex(vr_atom) do
    case vr_atom do
      :US -> "US"
      :UL -> "UL"
      :UI -> "UI"
      :LO -> "LO"
      :SH -> "SH"
      :CS -> "CS"
    end
  end

  def serialize_data_element(data_element, endianness: endianness, explicit: true) do
    group = serialize_u16(data_element.group_number, endianness)
    element = serialize_u16(data_element.element_number, endianness)                                                                                                                            

    # TODO cannot handle VM > 1
    value =
      case data_element.vr do
        :US -> serialize_u16(data_element |> DataElement.value(), endianness)
        :UL -> serialize_u32(data_element |> DataElement.value(), endianness)
        :UI -> serialize_uid(data_element |> DataElement.value(), endianness)
        :LO -> serialize_u16(data_element |> DataElement.value(), endianness)
        :SH -> serialize_sh(data_element |> DataElement.value(), endianness)
        :CS -> serialize_cs(data_element |> DataElement.value(), endianness)
      end
                                                                                                                                                                                                
    value_length = serialize_u32(byte_size(value) + 2, endianness)
    res = group <> element <> vr_to_hex(data_element.vr) <> value_length <> value
    res
  end
```

Then I created a PDU.serialize_data_pdu function that is a copy of PDU.serialize, but instead of returning this PDV:

`pdv = <<pdv_length::32, presentation_context_id::8, 3::8, data::binary>>`

I return this:

`pdv = <<pdv_length::32, presentation_context_id::8, 2::8, data::binary>>`

The only difference is the `2::8` instead of the `3::8.` That is a flag indicating `Data, Last Fragment `instead of  `Command, Last Fragment`. This is very, very important!.

leonardorame commented:
Of course, this is the first running draft. It needs some polishing, but I'm very happy with the results so far.

leonardorame commented:
Mmm, there's something wrong in the serialization routine. When I check with WireShark all values start with `\000\000`:

![image](https://github.com/user-attachments/assets/c2cfb9e2-c4f2-44f1-bf8a-53ecf8eab566)

leonardorame commented:
Fixed for strings. Numbers still adds `\000\000`.

![image](https://github.com/user-attachments/assets/030d9871-98fa-4102-897f-5509d4300425)

leonardorame commented:
I was wrongly serializing the LO vr assuming those were numbers, but are long strings. Fixed.

jjedele commented:
Hey @leonardorame ,
super cool - good progress!

Some advice: Don't test your serialization logic in Wireshark, use unit tests. These functions should be very nice to test and are just the opposite direction of the current parsing logic. You will speed up your development process a lot, and have automated tests that make sure you won't break it again. See example I linked above.

Otherwise I'll try to give some more feedback to your code tomorrow evening.

Question: Do you already have a goal for this PR in mind (i.e., in which state would you want to merge it)? I think implementing full serialization for all VRs is a bit too much to pack in this PR as well. I could imagine e.g. to clean this one up a bit, and add an API concept, so a user of this library could do at least something with it (instead of just always returning the same static responses).

leonardorame commented:
Hi @jjedele, my initial goal was to see how far can I get regarding this C-FIND-MWL implementation, now I'm thinking of polishing it more to be able to use it in a real application (I dream of having a dicom MWL that can be monitored real time using a LiveView application, for example). O course, I would like this to be added to the library.

I agree I must add tests, also I would like to discuss with you about the best way to implement an API for this.

Regarding the API, I simply added `register_cfind_getresponses` callback, to let the user decide how to return the list (or stream in the future) of responses (from a database query, a file, etc.), the "engine" just asks for them, but don't know anything about the user's implementation. Here's one example:

```elixir
defmodule Worklist do
  use Application

  def start(_type, _args) do
                {:ok, endpoint_pid} = GenServer.start_link(DicomNet.Endpoint, port: 4242)
                DicomNet.Endpoint.register_listener(endpoint_pid, self())
                DicomNet.Endpoint.register_cfind_getresponses(endpoint_pid, &get_responses/1)

                # send data set, e.g. with DCMTK: storescu localhost 4242 dataset.dcm
    loop()
    {:ok, endpoint_pid}
  end

  defp loop() do
     receive do
          {:dicom, %{operation: :cstore, dataset: ds}} ->
              IO.inspect("Received data via C-STORE.")
              # do something with the data set
          end
    loop()
  end;

  defp get_responses(dataset) do
    # Replace tags with atoms
    fields = Enum.map(dataset, fn {k, v} ->
          group = v.group_number
          element = v.element_number
          values = v.values
          case {group, element} do
              {0x0008, 0x0050} -> {:AccessionNumber, values}
              {0x0010, 0x0010} -> {:PatientName, values}
              {0x0010, 0x0020} -> {:PatientID, values}
            _ -> {:Ignore, nil}
          end
        end
    ) 

    # sample study list
    studies = [
      ["PatientName": "Carmack^John", PatientID: "1", AccessionNumber: "A001"],
      ["PatientName": "Kernighan^Brian", PatientID: "2", AccessionNumber: "A002"],
      ["PatientName": "Torvalds^Linus", PatientID: "3", AccessionNumber: "A003"],
      ["PatientName": "Van Rossum^Guido", PatientID: "4", AccessionNumber: "A004"],
      ["PatientName": "Valim^José", PatientID: "5", AccessionNumber: "A005"]
    ]
    
    # Return the list excluding those fields not present in fields list.
    Enum.map(studies, fn study ->
      Keyword.take(study, Keyword.keys(fields)) 
      |>Dicom.DataSet.from_keyword_list()
    end) 
  end
end   
```

jjedele commented:
I think we actually don't need much more than this.

I was considering for a bit if it should be a message based API like with the C-Store, but I think I prefer this solution because it's simpler. One of the reasons to separate stuff into processes and let them talk by messages is to isolate errors. But here the Association is already an process, so only the concrete network connection would crash if the callback has problem. And that is probably the expected behavior anyways.

Actually I'm considering if it would be better to make C-Store behave the same, and then just give the handlers to the constructor of the endpoint. Then it makes the API even simpler.

Something like this:
```
{:ok, endpoint_pid} = GenServer.start_link(DicomNet.Endpoint, port: 4242, handlers: [cstore: &handle_cstore/1, cfind: &handle_cfind/1])
```

One thing I would look into is using a [Stream](https://hexdocs.pm/elixir/Stream.html) instead of a list. The nice thing with DICOM C-Find is that it's already using a streaming model, so you can nicely push each response as you get it, in case it takes a bit of time to fetch them from somewhere.

My reason to start this library was also to maybe create some nice LiveView based PACS at some point :D Can I ask you what you're working on with the MWL?

jjedele commented on lib/dicom/binary_format.ex:
```diff
@@ -574,4 +609,18 @@ defmodule Dicom.BinaryFormat do
   end
+
+  def serialize_data_data_set(data_set) do
+    serialization_options = [endianness: :little, explicit: true]
```
I think this method can simply be called `serialize`, as it will be the normal serialization method. The `serialize_command_dataset` is a special case.

Also please already move the `options` into the function signature and hard-code them in the association for now. Then the API here is already "final".

leonardorame commented:
> I think we actually don't need much more than this.
> 
> I was considering for a bit if it should be a message based API like with the C-Store, but I think I prefer this solution because it's simpler. One of the reasons to separate stuff into processes and let them talk by messages is to isolate errors. But here the Association is already an process, so only the concrete network connection would crash if the callback has problem. And that is probably the expected behavior anyways.

I used the function handler instead of messages because I thought the later where async (I still don't know), and I needed a synchronous function. Maybe we could use both styles, depending on the use case.

> 
> Actually I'm considering if it would be better to make C-Store behave the same, and then just give the handlers to the constructor of the endpoint. Then it makes the API even simpler.
> 
> Something like this:
> 
> ```
> {:ok, endpoint_pid} = GenServer.start_link(DicomNet.Endpoint, port: 4242, handlers: [cstore: &handle_cstore/1, cfind: &handle_cfind/1])
> ```

Yes, I like the idea. What I don't know is what happens if, for example the cfind handler isn't declared but the user issues a CFind. Is there a way to only call it if it exists?, i.e. by checking if it's not null.

> One thing I would look into is using a [Stream](https://hexdocs.pm/elixir/Stream.html) instead of a list. The nice thing with DICOM C-Find is that it's already using a streaming model, so you can nicely push each response as you get it, in case it takes a bit of time to fetch them from somewhere.

Yes, of course. Streams will be a better approach than lists.

> My reason to start this library was also to maybe create some nice LiveView based PACS at some point :D Can I ask you what you're working on with the MWL?

At my company, since last year, we're migrating to Elixir, so far with great success (including the dicom viewer, based on LiveView with Cornerstone3D). We already migrated many not so complex apps. but leave the PACS for this part of the year as is the most complex application we have, it's based on C++ (dcmtk) for the dicom server, FreePascal for CGIs, HTML/CSS/JS for frontend and Python/Shell for scripting, but thanks to Elixir/Phoenix/LiveView we can build everything on one platform.

The MWL is part of out dicom server, based on DCMTK. But as is a Linux/Unix process without frontend we must ssh to the server to see the logs when something fails or to restart, reconfigure, etc. These tasks are not easy for junior tech support people and it takes a lot of time to teach them, so what I would like with a LiveView application is a way to "visually" handle those tasks.

jjedele commented:
> I used the function handler instead of messages because I thought the later where async (I still don't know), and I needed a synchronous function. Maybe we could use both styles, depending on the use case.

Genservers in Erlang/Elixir have a concept of synchronous request/reply messaging: https://hexdocs.pm/elixir/1.12/GenServer.html#call/3 . But then we would force the request handler to be a GenServer, which I don't think is necessary/a good idea.

> Yes, I like the idea. What I don't know is what happens if, for example the cfind handler isn't declared but the user issues a CFind. Is there a way to only call it if it exists?, i.e. by checking if it's not null.

Yes, something like this, and there's even a DICOM concept for this: [presentation context negotiation](https://dicom.nema.org/dicom/2013/output/chtml/part07/sect_D.3.html). Right now, [the association just by default accepts all presentation contexts](https://github.com/jjedele/dicom.ex/blob/main/lib/dicom_net/association.ex#L74) proposed by the client. I think a proper way of implementing this, would be as follows:

When you connect to server with a C-FIND SCU, it should propose a [special set of presentation contexts](https://github.com/pydicom/pynetdicom/blob/main/pynetdicom/sop_class.py#L269) for query/retrieve. Before accepting these, we could check if there is a c-find callback registered with the endpoint, which could handle such requests. Otherwise we decline and the findscu will stop, saying something along the lines that the contacted server doesn't it.

> At my company, since last year, we're migrating to Elixir, so far with great success 

That sounds super interesting. So far I unfortunately only managed to use Elixir in hobby projects like this, but I hope I can change that at some point. The whole stack with OTP/Phoenix/Ecto just seems so nice to work with. Almost everything feels like I would have put it on a wish list. Only hoping the typing story in Elixir will become a bit better, but it seems they are working hard on that.

leonardorame commented:
Great, I was thinking of starting with an implementation of rejection based on presentation contexts and/or callin/called AETITLEs (this by adding a new parameter to start_link apart from cstore and cfind, called policy for example). But before that, do you think this PR could be accepted in the current state?

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -199,4 +232,64 @@ defmodule DicomNet.Association do
+
+  defp cfind_response_data(presentation_context_id, response) do
+    Dicom.BinaryFormat.serialize(response, [endianness: :little, explicit: true])
+      |> Pdu.new_data_pdu(presentation_context_id, 0x02)
```
Between here and line 250 there is some asymmetry between the APIs: One time we set the flag here, one time it's set in Pdu via default. Result: Both modules know about some of this PDU type flags.

For me, it seems to be the `Pdu` module's job to know about how exactly these flags are encoded. The public API of Pdu (the one used here) should not force a user to know specific hex codes. I could imagine this to look rather like `Pdu.new_data_pdu(pc_id, last: true/false, type: :command / :data)`. And then Pdu decides on the hex value based on these arguments.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -199,4 +232,64 @@ defmodule DicomNet.Association do
+  end
+
+  defp cfind_response_data(presentation_context_id, response) do
+    Dicom.BinaryFormat.serialize(response, [endianness: :little, explicit: true])
```
Would add a `# TODO: options are hardcoded and need to be obtained from PC` comment here.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -199,4 +232,64 @@ defmodule DicomNet.Association do
+    # of Dicom.DataSet.from_keyword_list.
+    stream = state.getresponses.(data_set)
+    Enum.map(stream, fn response ->
+        headerbuffer = cfind_response_header(command, presentation_context_id, 0xff00)
```
And a bit nitpicky, but the `buffer` in these variable names is superfluous. Doesn't tell me more than just `header`, and also it's not really a buffer since everything in Elixir is immutable.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -199,4 +232,64 @@ defmodule DicomNet.Association do
+        :gen_tcp.send(socket, databuffer)
+    end)
+
+    response_tail = cfind_response_tail(asci_de, mid_de, 0x0000)
```
Now this is also not really nice, since the majority of the response is sent here, and the last part in another method. But I also need to think about how to clean this up, so OK for now.

jjedele reviewed (commented):
Left a review about the code, after that I would be fine with merging this :+1: 

Usually I would also say let's write unit tests for the serialization before, but I want to refactor them, and then we will get them basically for free. So that's OK for now.

leonardorame commented on lib/dicom_net/association.ex:
```diff
@@ -22,9 +22,10 @@ defmodule DicomNet.Association do
 
   @impl true
-  def init(%{socket: socket, event_listener: event_listener}) do
+  def init(%{socket: socket, event_listener: event_listener, getresponses: getresponses}) do
```
Yes, I like that idea.

jjedele commented on test/dicom_net/endpoint_integration_test.exs:
```diff
@@ -27,4 +27,52 @@ defmodule DicomNet.EndpointIntegrationTest do
+      ["PatientName": "Kernighan^Brian", PatientID: "2", AccessionNumber: "A002"],
+      ["PatientName": "Torvalds^Linus", PatientID: "3", AccessionNumber: "A003"],
+      ["PatientName": "Van Rossum^Guido", PatientID: "4", AccessionNumber: "A004"],
+      ["PatientName": "Valim^José", PatientID: "5", AccessionNumber: "A005"]
```
Strong selection ;) Let's hope they don't have anything serious!

leonardorame commented:
Hi @jjedele, I pushed a big commit with your suggestions. Could you check?

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -242,53 +296,57 @@ defmodule DicomNet.Association do
     Dicom.BinaryFormat.serialize_command_data_set(response)
-      |> Pdu.new_data_pdu(presentation_context_id) 
-      |> Pdu.serialize()
+    |> Pdu.new_data_pdu(presentation_context_id, type: :command_last_fragment)
```
Perfect! So much easier to read the code now, and directly see that command dataset is used with command PDU type.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -89,17 +98,31 @@ defmodule DicomNet.Association do
-      |> Map.put(:state, :association_established)
-      |> Map.put(:association, association_data)
+    case association_data.calling_ae_title do
+      "FINDSCU" ->
```
I assume this is coming from your other experiments and is not supposed to be in here?

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -160,44 +183,75 @@ defmodule DicomNet.Association do
+              |> Pdu.new_data_pdu(presentation_context_id, type: :command_last_fragment)
+              |> Pdu.serialize()
+
+            {:ok, cstore_Handler} ->
```
I'm being nitpicky here now - but please just lower_snake_case for variable names and atoms: `cstore_handler`. Modules are UpperCamelCase. Keeping the code style consistent is important :)

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -160,44 +183,75 @@ defmodule DicomNet.Association do
-            response_ds = handle_cstore(command, ds)
-            Dicom.BinaryFormat.serialize_command_data_set(response_ds)
-              |> Pdu.new_data_pdu(presentation_context_id)
+            {:ok, cfind_Handler} ->
```
`cfind_handler`

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -160,44 +183,75 @@ defmodule DicomNet.Association do
 
     # TODO this is not always c-store
+    ds = Dicom.BinaryFormat.from_binary(data, ts_options)
+    command_Code = DataSet.value_for!(command, :CommandField)
```
`command_code`

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -242,53 +296,57 @@ defmodule DicomNet.Association do
   end
 
-  defp handle_cfind(socket, command, data_set, presentation_context_id, ts_options, state) do
+  defp handle_cfind(socket, command, data_set, presentation_context_id, ts_options, cfind_Handler) do
```
`cfind_handler`

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -160,44 +183,75 @@ defmodule DicomNet.Association do
-            msg = {:dicom, %{operation: :cstore, dataset: ds}}
-            send(event_listener, msg)
+        _ ->
+          IO.inspect("command field not determied")
```
Should at this point rather be a `Logger.warning` than a `IO.inspect` I think

jjedele commented on lib/dicom_net/endpoint.ex:
```diff
@@ -72,8 +76,8 @@ defmodule DicomNet.Endpoint do
-  def handle_cast({:register_cfind_getresponses, function}, state) do
-    new_state = %{state | getresponses: function}
+  def handle_cast({:register_handlers, handlers}, state) do
+    new_state = %{state | handlers: handlers}
```
Is there a reason you left this message-based registration in the endpoint instead of also just making this a constructor argument? Don't really see a benefit in it anymore.

jjedele commented on lib/dicom_net/association.ex:
```diff
@@ -17,15 +17,24 @@ defmodule DicomNet.Association do
 
   @impl true
-  def init(%{socket: socket, event_listener: event_listener, getresponses: getresponses}) do
+  def init(%{socket: socket, event_listener: event_listener, handlers: handlers}) do
```
`event_listener` should now be removed from the APIs of both Endpoint and Association - it's not used anynmore, right?

jjedele reviewed (commented):
Thx - looking pretty good now. A little bit more cleanup and we can merge it :muscle:

leonardorame commented:
Hey @jjedele, please take a look now.

jjedele commented:
Looks great - tell me when you're ready and I'll merge :clap:

jjedele reviewed (approved):
Looks great now - good to merge :clap: 

When merging, please select to squash the commits.

jjedele commented:
Ah Ok, i thought you can do it when I approve the PR. I'll merge it then. Thx again for the contribution!
